slobrok/src/vespa/slobrok/server/sbenv.cpp

```diff
@@ -116,15 +116,23 @@ SBEnv::SBEnv(const ConfigShim &shim)
       _rpcsrvmap()
 {
     srandom(time(nullptr) ^ getpid());
-    _rpcsrvmap.proxy().registerListener(_globalVisibleHistory);
-    _rpcsrvmap.proxy().registerListener(_localRpcMonitorMap);
+    _localMonitorSubscription = MapSubscription::subscribe(_rpcsrvmap.proxy(), _globalVisibleHistory);
+    _consensusSubscription = MapSubscription::subscribe(_rpcsrvmap.proxy(), _localRpcMonitorMap);
+    _globalHistorySubscription = MapSubscription::subscribe(_localRpcMonitorMap.dispatcher(), _consensusMap);
     _rpcHooks.initRPC(getSupervisor());
 }
 
+void SBEnv::cleanup() {
+    auto pList = _exchanger.getPartnerList();
+    for (const auto & partner : pList) {
+        _exchanger.removePartner(partner);
+    }
+}
 
 SBEnv::~SBEnv()
 {
     getTransport()->WaitFinished();
+    cleanup();
```

Comment:
is the double cleanup needed? (I would expect that WaitFinished blocks forever in the cases where the other cleanup was not called...)
consider doing cleanup only here, but removing the WaitFinished call.

Reply:
we expect the transport thread(s) to be finished already when we get here, yes - can we assert on that?
cleanup() is removed again, it was added to ensure that MapListeners were unregistered before destructors.